build-tools/packages/build-tools/src/type-test-generator/typeTestUtils.ts

```diff
@@ -0,0 +1,289 @@
+ * @returns The mapping between item and its type
+ */
+export function typeDataFromFile(file: SourceFile): Map<string, TypeData> {
+	const typeData = new Map<string, TypeData>();
+	const exportedDeclarations = file.getExportedDeclarations();
+
+	for (const declarations of exportedDeclarations.values()) {
+		for (const dec of declarations) {
+			getNodeTypeData(dec).forEach((td) => {
+				const fullName = getFullTypeName(td);
+				if (typeData.has(fullName)) {
+					// This system does not properly handle overloads: instead it only keeps the last signature.
+					console.warn(`skipping overload for ${fullName}`);
+				}
+				typeData.set(fullName, td);
+			});
+		}
+	}
+	return typeData;
+}
+
+/**
+ * Initializes TypeScript projects for the current and previous package versions and loads specific source files.
+ * @param typeDefinitionFilePath - The path to the type definition file for the previous version.
+ * @param previousBasePath - The path to the root of the previous package version.
+ * @param previousPackageName - The name of the previous package version.
+ * @returns {{ currentFile: SourceFile, previousFile: SourceFile }} - The loaded source files for the current and previous versions.
+ */
+export function initializeProjectsAndLoadFiles(
+	typeDefinitionFilePath: string,
+	previousBasePath: string,
+	previousPackageName: string,
+) {
+	const currentFile = new Project({
+		skipFileDependencyResolution: true,
+		tsConfigFilePath: "tsconfig.json",
+	}).getSourceFileOrThrow("index.ts");
+
+	const previousTsConfigPath = path.join(previousBasePath, "tsconfig.json");
+	const project = new Project({
+		skipFileDependencyResolution: true,
+		tsConfigFilePath: existsSync(previousTsConfigPath) ? previousTsConfigPath : undefined,
+	});
+
+	let previousFile: SourceFile;
+	// Check for existence of alpha and add appropriate file
+	if (existsSync(typeDefinitionFilePath)) {
+		project.addSourceFilesAtPaths(typeDefinitionFilePath);
+		previousFile = project.getSourceFileOrThrow(`${previousPackageName}-alpha.d.ts`);
+		// Fall back to using .d.ts
+	} else {
+		project.addSourceFilesAtPaths(`${previousBasePath}/dist/**/*.d.ts`);
+		previousFile = project.getSourceFileOrThrow("index.d.ts");
+	}
+
+	return { currentFile, previousFile };
+}
+
+/**
+ * Generates compatibility test cases between the previous type definitions and the current type map.
+ * This function constructs test cases to validate forward and backward compatibility of types.
+ * @param previousData - array of type data from the previous file
+ * @param currentTypeMap - map containing curren type data
+ * @param packageObject - package.json object containing type validation settings
+ * @param testString - array to store generated test strings
+ * @returns - string array representing generated compatibility test cases
+ */
+export function generateCompatibilityTestCases(
+	previousData: TypeData[],
+	currentTypeMap: Map<string, TypeData>,
```

Comment:
Out of curiosity: why are the types of `currentTypeMap` and `previousData` different? I would have expected them to be the same.